File: src/application/runtime-events.ts
import { createHash, randomUUID } from "node:crypto";

export type RuntimeEventSeverity = "info" | "warn" | "error";

export interface RuntimeEvent {
  eventId: string;
  fingerprint: string;
  runId: string;
  code: string;
  severity: RuntimeEventSeverity;
  source: string;
  subject: string;
  occurredAt: string;
  seq: number;
  message: string;
  metrics?: Record<string, unknown> | undefined;
}

export interface RuntimeEventInput {
  runId: string;
  code: string;
  severity: RuntimeEventSeverity;
  source: string;
  subject: string;
  occurredAt: string;
  seq: number;
  message: string;
  metrics?: Record<string, unknown> | undefined;
}

export function createRuntimeEvent(input: RuntimeEventInput): RuntimeEvent {
  return {
    eventId: randomUUID(),
    fingerprint: createRuntimeEventFingerprint(input),
    runId: input.runId,
    code: input.code,
    severity: input.severity,
    source: input.source,
    subject: input.subject,
    occurredAt: input.occurredAt,
    seq: input.seq,
    message: input.message,
    metrics: input.metrics,
  };
}

export function createRuntimeEventFingerprint(input: Omit<RuntimeEventInput, "message" | "metrics">): string {
  const base = `${input.runId}|${input.occurredAt}|${input.code}|${input.source}|${input.subject}|${input.seq}`;
  return createHash("sha256").update(base).digest("hex");
}

export interface MutationAuditInput {
  runId: string;
  seq: number;
  actor: string;
  path: string;
  action: "set" | "delete";
  accepted: boolean;
  reason: string;
  occurredAt: string;
}

export function createMutationAuditEvent(input: MutationAuditInput): RuntimeEvent {
  const decision = input.accepted ? "accepted" : "rejected";
  return createRuntimeEvent({
    runId: input.runId,
    code: "RUN_STATE_MUTATION",
    severity: input.accepted ? "info" : "warn",
    source: "run-state",
    subject: `${input.actor}:${input.path}`,
    occurredAt: input.occurredAt,
    seq: input.seq,
    message: `${decision} ${input.action} on ${input.path}: ${input.reason}`,
    metrics: {
      actor: input.actor,
      path: input.path,
      action: input.action,
      accepted: input.accepted,
      reason: input.reason,
    },
  });
}

File: crates/rlm-core/src/application/execution/runtime_events.rs
use std::sync::Mutex;

use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};
use uuid::Uuid;

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum RuntimeEventSeverity {
    Info,
    Warn,
    Error,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RuntimeEvent {
    pub event_id: String,
    pub fingerprint: String,
    pub run_id: String,
    pub code: String,
    pub severity: RuntimeEventSeverity,
    pub source: String,
    pub subject: String,
    pub occurred_at: String,
    pub seq: u64,
    pub message: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metrics: Option<serde_json::Value>,
}

pub struct RuntimeEventInput {
    pub run_id: String,
    pub code: String,
    pub severity: RuntimeEventSeverity,
    pub source: String,
    pub subject: String,
    pub occurred_at: String,
    pub seq: u64,
    pub message: String,
    pub metrics: Option<serde_json::Value>,
}

pub fn create_runtime_event_fingerprint(input: &RuntimeEventInput) -> String {
    let base = format!(
        "{}|{}|{}|{}|{}|{}",
        input.run_id, input.occurred_at, input.code, input.source, input.subject, input.seq
    );
    let mut hasher = Sha256::new();
    hasher.update(base.as_bytes());
    format!("{:x}", hasher.finalize())
}

pub fn create_runtime_event(input: RuntimeEventInput) -> RuntimeEvent {
    let fingerprint = create_runtime_event_fingerprint(&input);
    RuntimeEvent {
        event_id: Uuid::new_v4().to_string(),
        fingerprint,
        run_id: input.run_id,
        code: input.code,
        severity: input.severity,
        source: input.source,
        subject: input.subject,
        occurred_at: input.occurred_at,
        seq: input.seq,
        message: input.message,
        metrics: input.metrics,
    }
}

pub fn runtime_event_occurred_at_now() -> String {
    time::OffsetDateTime::now_utc()
        .format(&time::format_description::well_known::Rfc3339)
        .unwrap_or_else(|_| "1970-01-01T00:00:00Z".into())
}

pub trait RuntimeEventSink: Send + Sync {
    fn emit(&self, event: RuntimeEvent) -> Result<(), String>;
}

pub struct NoopRuntimeEventSink;

impl RuntimeEventSink for NoopRuntimeEventSink {
    fn emit(&self, _event: RuntimeEvent) -> Result<(), String> {
        Ok(())
    }
}

pub struct InMemoryRuntimeEventStore {
    pub events: Mutex<Vec<RuntimeEvent>>,
}

impl Default for InMemoryRuntimeEventStore {
    fn default() -> Self {
        Self::new()
    }
}

impl InMemoryRuntimeEventStore {
    pub fn new() -> Self {
        Self {
            events: Mutex::new(Vec::new()),
        }
    }
}

impl RuntimeEventSink for InMemoryRuntimeEventStore {
    fn emit(&self, event: RuntimeEvent) -> Result<(), String> {
        self.events
            .lock()
            .map_err(|err| err.to_string())?
            .push(event);
        Ok(())
    }
}
